File: javascript/322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
 var coinChange = function(coins, amount) {
    let dp = Array(amount + 1).fill(amount + 1);
    dp[0] = 0;
    for (let i = 1; i < dp.length; i++) {
        for (let j = 0; j < coins.length; j++) {
            if (i - coins[j] >= 0) {
                dp[i] = Math.min(dp[i], 1 + dp[i - coins[j]]);
            }
        }
    }
    return dp[amount] === amount + 1 ? -1 : dp.pop();
};

File: golang/322-coin-change.go
package golang

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0
	for i := 1; i < amount+1; i++ {
		dp[i] = -1
	}
    
    for i := 1; i <= amount; i++ {
        for _, coin := range coins {
			if i-coin >= 0 && dp[i-coin] != -1 {
				if dp[i] == -1 {
					dp[i] = dp[i-coin] + 1
				} else {
					dp[i] = min(dp[i], 1 + dp[i-coin])
				}
			}
		}

	}
	return dp[amount]
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}